src/smart-escrow/SmartEscrow.sol

```diff
@@ -0,0 +1,165 @@
+
+/**
+ * @title SmartEscrow contract
+ * @dev Contract to handle payment of OP tokens over a period of vesting with
+ * the ability to terminate the contract.
+ */
+contract SmartEscrow is Ownable2Step {
+    event OPTransfered(uint256 amount, address indexed recipient);
+    event BeneficiaryOwnerUpdated(address indexed oldOwner, address indexed newOwner);
+    event BeneficiaryUpdated(address indexed oldBeneficiary, address indexed newBeneficiary);
+    event ContractTerminated();
+
+    IERC20 public constant OP_TOKEN = IERC20(0x4200000000000000000000000000000000000042);
+
+    address public beneficiaryOwner;
+    address public beneficiary;
+    uint256 public released;
+    uint64 public immutable start;
+    uint64 public immutable end;
+    uint64 public immutable vestingPeriod;
+    uint64 private immutable _initialTokens;
+    uint64 private immutable  _vestingEventTokens;
+    bool public contractTerminated;
+
+    /**
+     * @dev Set initial parameters.
+     * @param beneficiaryOwnerAddress Address which can update the beneficiary address.
+     * @param beneficiaryAddress Address which receives tokens that have vested.
+     * @param escrowOwner Address which can terminate the contract.
+     * @param startTimestamp Timestamp of the start of vesting period (or the cliff, if there is one).
+     * @param endTimestamp Timestamp of the end of the vesting period.
+     * @param vestingPeriodSeconds Period of time between each vesting event in seconds.
+     * @param numInitialTokens Number of OP tokens which vest at start time.
+     * @param numVestingEventTokens Number of OP tokens which vest upon each vesting event.
+     */
+    constructor(
+        address beneficiaryOwnerAddress,
+        address beneficiaryAddress,
+        address escrowOwner,
+        uint64 startTimestamp,
+        uint64 endTimestamp,
+        uint64 vestingPeriodSeconds,
+        uint64 numInitialTokens,
+        uint64 numVestingEventTokens
+    ) {
+        if (beneficiaryOwnerAddress == address(0) || beneficiaryAddress == address(0)) {
+            revert AddressIsZeroAddress();
+        }
+        if (startTimestamp == 0) revert StartTimeIsZero();
+        if (startTimestamp > endTimestamp) revert StartTimeAfterEndTime(startTimestamp, endTimestamp);
+        if (vestingPeriodSeconds == 0) revert VestingPeriodIsZeroSeconds();
+
+        beneficiary = beneficiaryAddress;
+        beneficiaryOwner = beneficiaryOwnerAddress;
+        start = startTimestamp;
+        end = endTimestamp;
+        vestingPeriod = vestingPeriodSeconds;
+        _initialTokens = numInitialTokens;
+        _vestingEventTokens = numVestingEventTokens;
+
+        _transferOwnership(escrowOwner);
+    }
+
+    /**
+     * @dev Allow escrow owner (2-of-2 multisig with beneficiary and benefactor
+     * owners as signers) to terminate the contract.
+     *
+     * Emits a {ContractTerminated} event.
+     */
+    function terminate(address returnAddress) external onlyOwner {
```

Comment:
these should be onlySuperAdmin or whatever? Or is do we control this?

Reply:
that's what `onlyOwner` should do?

Reply:
so OP needs us to agree to terminate? Seems ok